File: rust/src/2750.rs
fn main() {
    let mut buf = String::new();
    read_line(&mut buf);

    let n = parse_int(&buf);

    let mut arr: Vec<i32> = (0..n)
        .map(|_| {
            read_line(&mut buf);
            parse_int(&buf)
        })
        .collect();

    // bubble_sort(&mut arr);
    // selection_sort(&mut arr);
    // insertion_sort(&mut arr);
    quick_sort(&mut arr[..], n);

    for num in arr {
        println!("{num}");
    }
}

fn bubble_sort(arr: &mut Vec<i32>) {
    let len = arr.len();

    for i in 0..len - 1 {
        for j in 0..len - 1 - i {
            if arr[j] < arr[j + 1] {
                continue;
            }

            arr.swap(j, j + 1);
        }
    }
}

fn selection_sort(arr: &mut Vec<i32>) {
    let len = arr.len();

    for i in 0..len - 1 {
        let (min_index, _) = arr[i..].iter().enumerate().min_by_key(|(_, &n)| n).unwrap();

        arr.swap(i, min_index + i);
    }
}

fn insertion_sort(arr: &mut Vec<i32>) {
    let len = arr.len();

    for i in 1..len {
        for j in (0..=i - 1).rev() {
            if arr[j] < arr[j + 1] {
                break;
            }

            arr.swap(j, j + 1);
        }
    }
}

fn quick_sort(arr: &mut [i32], len: i32) {
    if len <= 1 {
        return;
    }

    let mut i = 0;
    let mut j = len - 1;
    let pivot_value = arr[(len / 2) as usize];

    while i <= j {
        while arr[i as usize] < pivot_value {
            i += 1;
        }
        while arr[j as usize] > pivot_value {
            j -= 1;
        }

        if i > j {
            break;
        }

        arr.swap(i as usize, j as usize);
        i += 1;
        j -= 1;
    }

    quick_sort(&mut arr[..=j as usize], j + 1);
    quick_sort(&mut arr[i as usize..], len - i);
}

fn read_line(buf: &mut String) {
    buf.clear();
    std::io::stdin().read_line(buf).unwrap();
}

fn parse_int(buf: &String) -> i32 {
    buf.trim().parse().unwrap()
}

File: python/2750.py
def main():
    N = int(input())
    arr = [int(input()) for _ in range(N)]

    # bubble_sort(arr)
    # selection_sort(arr)
    # insertion_sort(arr)
    quick_sort(arr, 0, N - 1)

    print(*arr, sep="\n")


def bubble_sort(arr):
    length = len(arr)

    for i in range(length - 1):
        for j in range(length - 1 - i):
            if arr[j] <= arr[j + 1]:
                continue

            arr[j], arr[j + 1] = arr[j + 1], arr[j]


def selection_sort(arr):
    for i in range(len(arr) - 1):
        right_part = arr[i:]
        min_index = right_part.index(min(right_part)) + i

        arr[i], arr[min_index] = arr[min_index], arr[i]


def insertion_sort(arr):
    for i in range(1, len(arr)):
        for j in range(i - 1, -1, -1):
            if arr[j] <= arr[j + 1]:
                break

            arr[j], arr[j + 1] = arr[j + 1], arr[j]


def quick_sort(arr, left, right):
    if left >= right:
        return

    i, j = left, right
    pivot = arr[(i + j) // 2]

    while i <= j:
        while arr[i] < pivot:
            i += 1

        while arr[j] > pivot:
            j -= 1

        if i > j:
            break

        arr[i], arr[j] = arr[j], arr[i]
        i += 1
        j -= 1

    quick_sort(arr, left, j)
    quick_sort(arr, i, right)


if __name__ == "__main__":
    main()
